### ISSUE TITLE
Deferred rendering

### ISSUE THREAD

**Issue by Tomash667** [Tomash667/carpg#131]
ABOUT SHADING:
When we talk about shading there are generally two approaches: forward rendering and deferred rendering.
Deferred shading is preferred technique because most of the complicated work is done only once (scene is rendered in g-buffers) and then you can easily render lights and post-process effects.
Regardless what approach you use, Phong shading for lighting calculation is a standard: N*L*diffuse*material + N*H*specular (L-inverse light direction; N-surface normal, H-halfway vector - for calculating reflection vector)
If you want HDR rendering, use higher precision floating point texture to accumulate (diffuse*material + specular) and then apply tone mapping.
If you don't want HDR rendering, use one regular texture for accumulating diffuse light (+ ambient) and other for specular light (+ emissive). At the end you have 3 textures and final color is: diffuse*material + specular. This way you get fake HDR effect - specular reflections will look great and you wont get that washed out effect.
Most people do lighting calculations wrong: (diffuse + specular)*material. 
This is wrong for two reasons: 
1. specular is reflective component and therefore it shouldn't be multiplied with material (just added)
2. if you have more lights, scene will look washed out because of color clamping. Sample is Ridick: Escape for Buttcher Bay - they did calculation all wrong and slow. If you had more that one light everything was white as if everything is reflective - washed out..
ABOUT SHADOWS:
FEAR and DOOM 3 use shadow volumes. It is a decent technique but very rarely used these days because of fill rate problem and inability to handle bit-masked geometry (vegetation, fences, ...) and on DX9 it requires too much CPU calculations.
Standard today is shadow mapping. For outdoor environments pretty much everyone uses Cascaded Shadow Maps (stable version where pixel centers are aligned and shadow map is not rotated). For indoor, everyone uses standard shadow mapping for spot lights and cube shadow mapping for small omnidirectional lights.
Problem with all these shadow techniques it that they are messy and can't handle all the cases. Cascaded SM is slow, requires much memory, has wastage problems, can't handle infinite environments and people actually bother with stabilizing them even though they are used in dynamic environment full of vegetation affected by the wind. Even smallest movement of geometry will ruin stabilization altogether (Crysis did it for some reason even though it was useless).
Cube shadow mapping is limitating because you can only handle small omni lights and special cube texture is required.
